File: scripts/FigureScripts/Gains_Of_Aggregation.jl


# --- Project Modules ---
include("../common_setup.jl")

# --- External Packages ---
using Plots, Dates, Combinatorics, Serialization
GC.gc() # Run garbage collection to free memory, useful for repeat runs

plot_data = deserialize(joinpath(@__DIR__, "..", "..", "results", "cache", "17ClientWeekly.jls"))

# =========================
# 1. Load data from serialized file
# =========================
println("Loading data from serialized file")

# Extract clients and coalition costs directly from plot_data
clients = plot_data.clients
coalition_costs = plot_data.coalition_costs

println("Number of clients: ", length(clients))
println("Number of coalitions: ", length(coalition_costs))

# average_cost_ratio[i] = unweighted average cost ratio for coalitions of size i (not per-client)
average_cost_ratio = zeros(Float64, length(clients))

# Store all individual cost ratios for plotting
all_cost_ratios = Dict{Int, Vector{Float64}}()

for i in 1:(length(clients))
    # Calculate the unweighted average cost ratio for coalitions of size i
    coalitions_of_size_i = [coalition for coalition in keys(coalition_costs) if length(coalition) == i]
    cost_ratios = Float64[]

    for coalition in coalitions_of_size_i
        # Calculate cost ratio for this coalition
        coalition_cost = coalition_costs[coalition]
        singleton_sum = sum(coalition_costs[[client]] for client in coalition)
        push!(cost_ratios, coalition_cost / singleton_sum)
    end

    # Store all cost ratios for this coalition size
    all_cost_ratios[i] = cost_ratios

    # Store the unweighted average cost ratio for coalitions of size i
    average_cost_ratio[i] = mean(cost_ratios)
end

# First, create plot with filled area showing range of coalition costs
p = plot(; xlabel="Number of Clients in Coalition",
         ylabel="Relative cost [%]",
         xticks=1:(length(clients)),
         xrotation = 45,
         #legend=:topright,
         SMALL_FIGURE_STYLE...,
         xguidefont=font(8, "Times Roman"),
         legendfontsize=6)

# Calculate min and max for each coalition size to create filled area
x_vals = Int[]
min_vals = Float64[]
max_vals = Float64[]

for i in 1:(length(clients))
    if haskey(all_cost_ratios, i) && !isempty(all_cost_ratios[i])
        push!(x_vals, i)
        push!(min_vals, minimum(all_cost_ratios[i]) * 100)
        push!(max_vals, maximum(all_cost_ratios[i]) * 100)
    end
end

# Create filled area by plotting upper and lower bounds
plot!(p, x_vals, min_vals,
      fillrange=max_vals,
      fillalpha=0.8,
      fillcolor=PALETTE_SAND,
      linewidth=0,
      color=:transparent,
      label="Relative cost range")

# Then add the averaged values on top
plot!(p, 1:(length(clients)), average_cost_ratio[1:(length(clients))]*100,
      marker=:circle,
      markersize=3,
      linewidth=2,
      color=PALETTE_LIGHT_GREEN,
      label="Unweighted average")

display(p)
savefig(p, "Gains_of_Aggregation_17Clients_New.png")

File: python/figures/gains_of_aggregation.py
"""Port of scripts/FigureScripts/Gains_Of_Aggregation.jl."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import matplotlib.pyplot as plt

import style


def plot_gains_of_aggregation(coalition_costs, clients):
    """`coalition_costs` is a dict keyed by frozenset[str] -> float (see data_io.PlotData).
    Returns a matplotlib Figure."""
    n = len(clients)
    average_cost_ratio = [None] * n
    all_cost_ratios: dict[int, list[float]] = {}

    for i in range(1, n + 1):
        coalitions_of_size_i = [c for c in coalition_costs if len(c) == i]
        cost_ratios = []
        for coalition in coalitions_of_size_i:
            coalition_cost = coalition_costs[coalition]
            singleton_sum = sum(coalition_costs[frozenset([c])] for c in coalition)
            cost_ratios.append(coalition_cost / singleton_sum)
        all_cost_ratios[i] = cost_ratios
        average_cost_ratio[i - 1] = sum(cost_ratios) / len(cost_ratios)

    x_vals, min_vals, max_vals = [], [], []
    for i in range(1, n + 1):
        if all_cost_ratios.get(i):
            x_vals.append(i)
            min_vals.append(min(all_cost_ratios[i]) * 100)
            max_vals.append(max(all_cost_ratios[i]) * 100)

    fig, ax = plt.subplots()
    ax.fill_between(
        x_vals, min_vals, max_vals,
        color=style.PALETTE["sand"], alpha=0.8, linewidth=0,
        label="Relative cost range",
    )
    ax.plot(
        range(1, n + 1), [v * 100 for v in average_cost_ratio],
        marker="o", markersize=3, linewidth=2,
        color=style.PALETTE["light_green"], label="Unweighted average",
    )
    ax.set_xlabel("Number of Clients in Coalition")
    ax.set_ylabel("Relative cost [%]")
    ax.set_xticks(range(1, n + 1))
    plt.setp(ax.get_xticklabels(), rotation=45)
    ax.legend()
    fig.tight_layout()

    return fig
